### PR TITLE
Implement exp/log/power SQL functions

### LINKED ISSUE #5812: sql: support log/exp
Add support for the log/exp operators https://www.postgresql.org/docs/13/functions-math.html

These are likely needed for histogram support, though that hasn't been specced out yet. Good to support either way!

### PR DESCRIPTION
**Pull request by antifuchs**
Fixes #5812, #5329

In this PR, we implement `log` (unary and binary), as well as `power` and the unary `ln`, `log10` and `exp` SQL functions.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5885)

### CODE REVIEW

File: src/expr/src/scalar/func.rs
Comment (ruchirK) [inline src/expr/src/scalar/func.rs:3386]: some other functions on decimals have this scale check
```
          CeilDecimal(scale) | FloorDecimal(scale) | RoundDecimal(scale) | SqrtDec(scale) => {
                match input_type.scalar_type {
                    ScalarType::Decimal(_, s) => assert_eq!(*scale, s),
                    _ => unreachable!(),
                }
                input_type.scalar_type.nullable(in_nullable)
            }
```

not sure if thats also needed here / if we can slot these functions into that match arm

**Review by ruchirK — APPROVED** [review]
Other than the one comment these changes look good to me. Sidenote: it might be interesting to run the underlying code against quickcheck or something to see if we ever fail any `exp ( ln (x)) == x` invariants -- I don't feel confident in my understanding of f64 to have an intuition for what kinds of `x` would fail that at the moment

**Comment by antifuchs** [comment]
quickcheck sounds very good - I'll work on these two now.

File: src/expr/src/scalar/func.rs
Comment (ruchirK) [inline src/expr/src/scalar/func.rs:1291]: we are following the existing style of the code here so not a huge deal but I'm a little bit weirded out by how many datum conversions we are doing here -- like if I'm reading it right:

line 1291 unwraps a Datum::Decimal and converts it to a Datum::Float64
line 1293 unwraps the Datum::Float64 into a regular f64
lines 1296 and 1297 create two new datums a Datum::Float64 and a Datum::Int32, and then finally
line 1295 converts those into a Datum::Decimal

I didn't look at the generated assembly / don't think this is a big performance deal at this point but I feel like this all makes the code challenging to understand for a first time user. not totally sure why we write these functions like this my guess is that we already have these Datum::Type1 -> Datum::Type2 casts for the actual casting and then its convenient to reuse them here? is there another reason why we are doing things this way?

Comment (ruchirK) [reply src/expr/src/scalar/func.rs:1291]: @antifuchs this is me just musing about the code / trying to understand it better. I think the code is all correct as written and we should merge it like this once it passes tests

**Comment by ruchirK** [comment]
> quickcheck sounds very good - I'll work on these two now.

the quickcheck thing was just me musing again! I think its a neat thing to check out but I don't think it needs to block merging these changes or anything like that

File: src/expr/src/scalar/func.rs
Comment (antifuchs) [reply src/expr/src/scalar/func.rs:1291]: So, I think you're right in feeling weird about this. As we see in the failing test, there's something sketch happening in decimals, and it may very well be rooted in the type conversions. I'll see if behavior changes if we change the type conversion behavior a bit.

File: doc/user/data/sql_funcs.yml
Comment (JLDLaughlin) [inline doc/user/data/sql_funcs.yml:110]: 🎉 

yay docs!

File: src/expr/src/scalar/func.rs
Comment (JLDLaughlin) [inline src/expr/src/scalar/func.rs:1247]: ooh nice!

File: src/expr/src/scalar/func.rs
Comment (JLDLaughlin) [inline src/expr/src/scalar/func.rs:1258]: tiny, ignorable nit: the `i32::from(scale)` bit is separated out in the other functions you've added, if you'd want to make it a separate line here as well

**Review by JLDLaughlin — COMMENTED** [review]
these changes look great to me! it might be worth pulling @philip-stoev in to see if he has any additional test cases for you.

File: src/expr/src/scalar/func.rs
Comment (antifuchs) [reply src/expr/src/scalar/func.rs:1258]: True! I had considered that but would have to re-convert to a `u8` again, or retain the original; I thought this one read better (:

File: test/sqllogictest/funcs.slt
Comment (sploiselle) [inline test/sqllogictest/funcs.slt:985]: can we get tests for zeroes and negative numbers for both of the inputs here, even if they're expected to error?

**Comment by benesch** [comment]
One quick thought: the decimal stuff is about to all get ripped out and replaced with an implementation that works much better, courtesy of @sploiselle. If we're not confident about any of the decimal functions here, we can just disable 'em for now (basically just comment out the line in the SQL func mod; might as leave the rest of the infrastructure in place).

**Comment by antifuchs** [comment]
> If we're not confident about any of the decimal functions here, we can just disable 'em for now (basically just comment out the line in the SQL func mod; might as leave the rest of the infrastructure in place).

With update to the test I pushed about 1/3 of the way to the end up there, I think I'm confident enough; they're not as precise as the postgresql functions, but they're not egregiously wrong. I'll let @sploiselle decide, but I think they're not worse off in terms of accuracy than other SQL functions we're defining, and would benefit as much as the other functions from his work (:

**Comment by sploiselle** [comment]
If you have it working, you've already done the hard part! If you want to bail on it for any reason, though, feel free to just comment them out as @benesch suggested.

**Review by sploiselle — APPROVED** [review]

**Pull request merged**
